dingo/preprocess.py

```diff
@@ -0,0 +1,245 @@
+        from the model. If this simultaneous removal produces an infesible
+        solution (or a solution of 0) to the objective function, 
+        these reactions are restored with their initial bounds.
+        
+        A dingo-type tuple is then created from the cobra model 
+        using the "cobra_dingo_tuple" function.
+        
+        The outputs are
+        (a) A list of the removed reactions ids
+        (b) A reduced dingo model
+        """        
+        
+        # create a list from the combined blocked, zero-flux, mle reactions
+        blocked_mle_zero = self._blocked_reactions + self._mle_reactions + self._zero_flux_reactions
+        list_removed_reactions = list(set(blocked_mle_zero))        
+        self._removed_reactions = list_removed_reactions
+   
+        # remove these reactions from the model
+        self._remove_model_reactions()
+                     
+        remained_reactions = list((Counter(self._initial_reactions)-Counter(self._removed_reactions)).elements())
+        remained_reactions = list((Counter(remained_reactions)-Counter(self._essential_reactions)).elements())
+        
+        tol = 1e-6
+        
+        if extend != 0 and extend != 1:
+            raise Exception("Wrong Input to extend parameter")
+  
+        # find additional reactions with a possibility of removal
+        additional_removed_reactions_count = 0       
+        for reaction in remained_reactions:
+            
+            fba_solution_before = self._model.optimize().objective_value
+                        
+            initial_lower = self._model.reactions.get_by_id(reaction).lower_bound
+            initial_upper = self._model.reactions.get_by_id(reaction).upper_bound
+            
+            # perform a knock-out and check the output
+            self._model.reactions.get_by_id(reaction).lower_bound = 0
+            self._model.reactions.get_by_id(reaction).upper_bound = 0
+    
+            fba_solution_after = self._model.optimize().objective_value
+            
+            if fba_solution_after != None and (extend == 1):
+                if (abs(fba_solution_after - fba_solution_before) < tol):
+                    self._removed_reactions.append(reaction)
+                    additional_removed_reactions_count += 1
+              
+            self._model.reactions.get_by_id(reaction).upper_bound = initial_upper
+            self._model.reactions.get_by_id(reaction).lower_bound = initial_lower
+            
+            
+        # compare FBA solution before and after the removal of additional reactions
+        fba_solution_initial = self._model.optimize().objective_value
+        self._remove_model_reactions()        
+        fba_solution_final = self._model.optimize().objective_value
+
+        
+        # get a list of the additional reactions to be removed
+        additional_removed_reactions_list = (self._removed_reactions[len(self._removed_reactions)-additional_removed_reactions_count:])
```

Comment:
instead of that complex line, you could keep a second list and add the reaction names after `additional_removed_reactions_count += 1`